src/java.base/share/classes/java/util/UUID.java

```diff
@@ -99,24 +104,190 @@ public final class UUID implements java.io.Serializable, Comparable<UUID> {
+                    UUID_COUNT = Integer.getInteger(PROP_NAME_UUID_COUNT, 256);
+                    if (UUID_COUNT < 1 || UUID_COUNT > Integer.MAX_VALUE / UUID_CHUNK) {
+                        throw new IllegalArgumentException(PROP_NAME_UUID_COUNT + " is out of range");
+                    }
+                    BUF_SIZE = UUID_CHUNK * UUID_COUNT;
+                } catch (Exception e) {
+                    throw new InternalError(e);
+                }
+            }
+
+            final SecureRandom random;
+            final StampedLock lock;
+            final byte[] buf;
+            int pos;
+
+            public Buffer(SecureRandom random) {
+                this.random = random;
+                this.lock = new StampedLock();
+                this.buf = new byte[BUF_SIZE];
+                this.pos = BUF_SIZE; // trigger re-creation on first use
+            }
+
+            public UUID next() {
+                long stamp = lock.tryOptimisticRead();
+                try {
+                    UUID uuid = null;
+
+                    // Optimistic path: optimistic locking succeeded.
+                    // Try to pull the UUID from the current buffer at current position.
+                    if (stamp != 0) {
+                        int p = (int)VH_POS.getAndAdd(this, UUID_CHUNK);
+                        if (p < BUF_SIZE) {
+                            uuid = new UUID(buf, p);
+                            if (lock.validate(stamp)) {
+                                // Success: UUID is valid, and there were no buffer changes.
+                                return uuid;
+                            }
+                        }
+                    }
+
+                    // Semi-pessimistic path: either the buffer was depleted, or optimistic locking
+                    // failed. Either way, we need to take the exclusive lock and try again.
+                    stamp = lock.tryConvertToWriteLock(stamp);
+                    if (stamp == 0L) {
+                        stamp = lock.writeLock();
+                    }
+
+                    // See if some other thread had already replenished the buffer.
+                    // Pull the UUID from there then. We are still holding the write lock, so
+                    // buffer is guaranteed to not change under our feet.
+                    if ((int)VH_POS.get(this) > 0) {
+                        int p = (int)VH_POS.getAndAdd(this, UUID_CHUNK);
+                        if (p < BUF_SIZE) {
+                            return new UUID(buf, p);
+                        }
+                    }
+
+                    // Pessimistic path: buffer requires replenishment. Recreate it from the
+                    // provided random, and initialize all UUIDs at once to avoid further
+                    // initializations, and thus false sharing between reader threads.
+                    random.nextBytes(buf);
+                    for (int c = 0; c < BUF_SIZE; c += UUID_CHUNK) {
```

Comment:
I think this could be faster by using a ByteBuffer (or VarHandle) to process as longs.

https://mail.openjdk.org/pipermail/core-libs-dev/2023-March/101249.html

Reply:
Yup, but that would probably fold into discussion here: https://github.com/openjdk/jdk/pull/14135#discussion_r1206097261